File: keyboards/ydkb/athena75_rgb_advanced/tools/host/synctime.c
// Copyright 2026 YANG
// SPDX-License-Identifier: GPL-2.0-or-later
//
// synctime: push the PC's wall-clock time to the Athena75 RGB over USB (raw-HID
// 0xFD 0x5E HH MM SS) so the MATRIX rain can show its HH:MM watermark. The board
// has no RTC: it free-runs the clock off its own timer once synced and loses it on
// power off, so run this on connect (and optionally on a loop to correct drift).
// Native only (Windows SetupAPI+hid.dll, macOS IOKit); local time via libc.
//
// Usage:  host_tool synctime [--utc] [--loop SEC]

#include "cmds.h"
#include "hid.h"
#include "proto.h"
#include "sys.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static int send_time(hid_dev *d, int use_utc) {
    time_t now = time(NULL);
    struct tm tmv;
#if defined(_WIN32)
    if (use_utc) gmtime_s(&tmv, &now); else localtime_s(&tmv, &now);
#else
    if (use_utc) gmtime_r(&now, &tmv); else localtime_r(&now, &tmv);
#endif
    uint8_t rep[ATHENA_REPORT_LEN] = {0};
    rep[0] = ATHENA_CMD; rep[1] = ATHENA_CLK_CMD;
    rep[2] = (uint8_t)tmv.tm_hour; rep[3] = (uint8_t)tmv.tm_min; rep[4] = (uint8_t)tmv.tm_sec;
    if (hid_write(d, rep) != 0) return -1;
    printf(">> synced %02d:%02d:%02d%s\n", tmv.tm_hour, tmv.tm_min, tmv.tm_sec, use_utc ? " UTC" : "");
    return 0;
}

int cmd_synctime(int argc, char **argv) {
    int use_utc = 0, loop = 0;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--utc")) use_utc = 1;
        else if (!strcmp(argv[i], "--loop") && i + 1 < argc) loop = atoi(argv[++i]);
        else { printf("unknown arg: %s\n", argv[i]); return 2; }
    }

    hid_dev *d = hid_open(ATHENA_VID, ATHENA_PID, ATHENA_USAGE_PAGE, ATHENA_USAGE);
    if (!d) { printf("error: device %04x:%04x not found\n", ATHENA_VID, ATHENA_PID); return 1; }

    int rc = send_time(d, use_utc);
    while (rc == 0 && loop > 0) {
        sys_msleep(loop * 1000);
        rc = send_time(d, use_utc);
    }
    hid_close(d);
    return rc == 0 ? 0 : 1;
}

File: keyboards/ydkb/athena75_rgb_advanced/tools/synctime.py
#!/usr/bin/env python3
# Copyright 2026 YANG
# SPDX-License-Identifier: GPL-2.0-or-later
#
# Push the PC's wall-clock time to the Athena75 RGB over USB (raw HID) so the
# MATRIX rain can show a dimmed HH:MM watermark. The board has no battery-backed
# RTC: it free-runs the clock off its own timer once synced, and loses the time
# on power off — so run this on connect (and optionally on a loop to correct
# drift). The firmware command is 0xFD 0x5E HH MM SS.
#
# Deps:  pip install hidapi
# Usage: python synctime.py                 # sync once (local time)
#        python synctime.py --loop 300       # re-sync every 300s
#        python synctime.py --utc            # send UTC instead of local time

import argparse
import sys
import time
from datetime import datetime

try:
    import hid
except ImportError:
    sys.exit("error: pip install hidapi")

DEFAULT_VID = 0x9D5B
DEFAULT_PID = 0x2514
USAGE_PAGE  = 0xFF60  # VIA raw-HID interface
USAGE       = 0x61
REPORT_LEN  = 32
CLK_CMD     = 0x5E


def find_path(vid, pid):
    """Pick the VIA raw-HID interface (usage page 0xFF60 / usage 0x61)."""
    for d in hid.enumerate(vid, pid):
        if d.get("usage_page") == USAGE_PAGE and d.get("usage") == USAGE:
            return d["path"]
    devs = hid.enumerate(vid, pid)  # fallback: some platforms omit usage
    return devs[0]["path"] if devs else None


def send_time(dev, use_utc):
    now = datetime.utcnow() if use_utc else datetime.now()
    payload = bytearray(REPORT_LEN)
    payload[:5] = bytes([0xFD, CLK_CMD, now.hour, now.minute, now.second])
    dev.write(b"\x00" + bytes(payload))  # leading 0 = report id
    print(f">> synced {now:%H:%M:%S}{' UTC' if use_utc else ''}")


def main():
    ap = argparse.ArgumentParser(description="Sync PC time to the Athena75 RGB MATRIX clock.")
    ap.add_argument("--loop", type=int, metavar="SEC", default=0,
                    help="keep running, re-syncing every SEC seconds (0 = once)")
    ap.add_argument("--utc", action="store_true", help="send UTC instead of local time")
    ap.add_argument("--vid", type=lambda s: int(s, 0), default=DEFAULT_VID)
    ap.add_argument("--pid", type=lambda s: int(s, 0), default=DEFAULT_PID)
    args = ap.parse_args()

    path = find_path(args.vid, args.pid)
    if not path:
        sys.exit(f"error: device {args.vid:#06x}:{args.pid:#06x} not found")

    dev = hid.device()
    dev.open_path(path)
    try:
        send_time(dev, args.utc)
        while args.loop > 0:
            time.sleep(args.loop)
            send_time(dev, args.utc)
    finally:
        dev.close()


if __name__ == "__main__":
    main()
